### DISCUSSION TITLE
@click.outside problem if you're using classes instead of x-show

### DISCUSSION THREAD

**Discussion by mokhosh** [alpinejs/alpine#2264] [7. Bugs (archived)] [answered]
This problem is not there in alpine 2, and i think the reason is this:

> It's worth noting that the .outside expression will only be evaluated when the element it's registered on is visible on the page. Otherwise, there would be nasty race conditions where clicking the "Toggle" button would also fire the @click.outside handler when it is not visible.

I think this is introduced in alpine 3, since i had the opposite problem in alpine two which led me to create this issue https://github.com/alpinejs/alpine/issues/207

So here is the problem. This is a common pattern:
```html
<div x-data="{open: false}">
  <button x-on:click="open = true">open</button>
  <div x-show="open" x-on:click.outside="open = false">modal</div>
</div>
```
This works perfectly in alpine 3. In alpine 2 you would have to say `x-on:click.away.prevent` to prevent "away" from being fired when the modal is not visible. I think @calebporzio has changed this in alpine 3 to prevent using prevent. But now the problem is you can't do this:
```html
<div x-data="{open: false}">
  <button x-on:click="open = true">open</button>
  <div x-show="open" x-bind:class="{hidden: ! open}">modal</div>
</div>
```
Which is a very useful pattern when creating responsive designs with tailwindcss.

**Comment by SimoTod** [comment, marked as answer]
What is the issue exactly? Sorry, it's not super clear. Your last example doesn't use click away. It also use show AND bind an hidden class at the same time which from a first glance looks redundant and a bit confusing to me.
Note that the click away not triggering when the element wasn't showing was also in Alpine v2. It hasn't been introduced in v3.

**Reply by mokhosh** [reply]
Sorry I substituted the wrong part while creating the example, this is the code that doesn't work as expected:
```html
<div x-data="{open: false}">
  <button x-on:click="open = true">open</button>
  <div x-bind:class="{hidden: ! open}" x-on:click.outside="open = false">modal</div>
</div>
```

**Reply by mokhosh** [reply]
> Note that the click away not triggering when the element wasn't showing was also in Alpine v2. It hasn't been introduced in v3.

Then I'm wrong about the cause of the issue. But the issue is definitely there. It works by downgrading to v2 and stops working by upgrading to v3.

**Reply by mokhosh** [reply]
To further clarify why I'm using a tailwind `hidden` class instead of `x-show`, I want the div to only care about `open` up to a certain screen breakpoint (mobile and tablet), and it will have a `lg:static` to make it always visible regardless of `open` property in desktop mode.
But that doesn't seem to work on v3.

**Reply by SimoTod** [reply]
I think v2 worked by coincidence. 
X-show delays the show phases one tick so when your event bubbles to the top, the element is still hidden. If you use bind, that doesn't happen and the class is applied before the event hits window so it hides it again.
As per v2, you can still use the `.stop` modifier on the button element.

**Reply by mokhosh** [reply]
I was ok in two. I need it to work on v3.

**Reply by SimoTod** [reply]
I gave you the solution 🤔 (which is the same solution mentioned in the ticket you opened for v2). On both v3 and v2, when you have this kind of issues, use the stop modifier. 

`<button x-on:click.stop="open = true">`

**Reply by mokhosh** [reply]
I'm sorry I misunderstood your answer. You're right. I vaguely remembered what happened back then and tried this:
```html
<div x-bind:class="{hidden: ! open}" x-on:click.stop.outside="open = false">modal</div>
```
which didn't work, but you meant I should add `.stop` to the main trigger button.

Thanks!

**Reply by SimoTod** [reply]
No worries

**Reply by beitomartinez** [reply]
Months later... but was looking exactly for this... `.stop` did the job!

**Comment by mokhosh** [comment]
Here's a full copy pastable example. Just try commenting v2 and uncommenting v3 to see the issue:
```html
<!DOCTYPE html>
<html>
<head>
    <!-- <script src="https://cdn.jsdelivr.net/gh/alpinejs/alpine@v2.x.x/dist/alpine.min.js" defer></script> -->
    <script defer src="https://unpkg.com/alpinejs@3.x.x/dist/cdn.min.js"></script>
    <link href="https://unpkg.com/tailwindcss@^2/dist/tailwind.min.css" rel="stylesheet">
</head>
<body>
    <div x-data="{open: false}">
        <button x-on:click="open = true">works on v2 only</button>
        <div x-bind:class="{hidden: ! open}" x-on:click.away="open = false">modal</div>
    </div>
    <div x-data="{open: false}">
        <button x-on:click="open = true">works on v2 and v3</button>
        <div x-show="open" x-on:click.away="open = false">modal</div>
    </div>
</body>
</html>
```

**Reply by MdAnowarHosen** [reply]
x-on:click.away="open = false" working well

**Reply by mokhosh** [reply]
Not sure what you mean, but you can copy paste the above example in a codepen and try it for yourself.

Also, this is a resolved discussion from 2021!
